File: core/src/net/artux/pda/map/ui/BackpackMenu.java
package net.artux.pda.map.ui;

import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.Touchable;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.VerticalGroup;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;

import net.artux.engine.utils.LocaleBundle;
import net.artux.pda.map.DataRepository;
import net.artux.pda.map.di.scope.PerGameMap;
import net.artux.pda.map.engine.AssetsFinder;
import net.artux.pda.map.engine.systems.SoundsSystem;
import net.artux.pda.map.engine.systems.player.PlayerSystem;
import net.artux.pda.map.ui.blocks.SlotTextButton;
import net.artux.pda.map.ui.view.DetailItemView;
import net.artux.pda.map.ui.view.HUD;
import net.artux.pda.map.ui.view.ItemsTableView;
import net.artux.pda.map.ui.view.OnItemClickListener;
import net.artux.pda.map.ui.view.bars.Utils;
import net.artux.pda.map.utils.Colors;
import net.artux.pda.map.utils.PlatformInterface;
import net.artux.pda.model.items.ArmorModel;
import net.artux.pda.model.items.ItemModel;
import net.artux.pda.model.items.ItemType;
import net.artux.pda.model.items.MedicineModel;
import net.artux.pda.model.items.WeaponModel;
import net.artux.pda.model.items.WearableModel;
import net.artux.pda.model.quest.story.StoryDataModel;

import java.util.List;

import javax.inject.Inject;

@PerGameMap
public class BackpackMenu extends Table {

    private final DataRepository dataRepository;
    private final AssetManager assetManager;
    private final FontManager fontManager;

    private final Label infoLabel;
    private final LocaleBundle localeBundle;
    private final ItemsTableView mainItemsView;

    private final DetailItemView armorView;
    private final DetailItemView rifleView;
    private final DetailItemView pistolView;

    @Inject
    public BackpackMenu(UserInterface userInterface, PlatformInterface platformInterface, HUD hud, SlotTextButton textButton,
                        SlotTextButton adButton, AssetsFinder assetsFinder, PlayerSystem playerSystem,
                        LocaleBundle localeBundle, DataRepository dataRepository, Skin skin, SoundsSystem soundsSystem) {
        super();
        this.localeBundle = localeBundle;
        this.dataRepository = dataRepository;
        this.assetManager = assetsFinder.getManager();
        top();
        setFillParent(true);
        defaults()
                .pad(10)
                .space(20);

        defaults().fill();


        fontManager = assetsFinder.getFontManager();
        Label.LabelStyle titleLabelStyle = fontManager.getLabelStyle(38, Color.WHITE);
        Label.LabelStyle subtitleStyle = fontManager.getLabelStyle(30, Color.GRAY);

        Table leftTable = new Table();
        leftTable.top().left();

        textButton.setText(localeBundle.get("main.close"));
        textButton.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                super.clicked(event, x, y);
                remove();
            }
        });

        adButton.setText(localeBundle.get("main.ad.rewarded"));
        adButton.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                platformInterface.rewardedAd();
            }
        });

        infoLabel = new Label("", subtitleStyle);

        VerticalGroup verticalGroup = new VerticalGroup();

        Table buttonsTable = new Table();
        buttonsTable.center();
        buttonsTable.defaults().center().space(20);
        buttonsTable.add(textButton).uniformX();
        buttonsTable.add(adButton).uniformX();
        verticalGroup.addActor(buttonsTable);

        buttonsTable = new Table();
        buttonsTable.center();
        buttonsTable.defaults().space(20).center();
        buttonsTable.add(hud);
        buttonsTable.add(infoLabel);
        buttonsTable.row();


        verticalGroup.addActor(buttonsTable);

        leftTable.add(verticalGroup)
                .left().top().row();

        armorView = new DetailItemView(null, titleLabelStyle, subtitleStyle, localeBundle, assetManager);
        leftTable.add(armorView)
                .fill();

        Table horizontalGroup = new Table();
        horizontalGroup.center();
        rifleView = new DetailItemView(null, titleLabelStyle, subtitleStyle, localeBundle, assetManager);
        pistolView = new DetailItemView(null, titleLabelStyle, subtitleStyle, localeBundle, assetManager);
        pistolView.disableDesc();
        rifleView.disableDesc();
        armorView.disableDesc();
        horizontalGroup.add(rifleView).fill().uniform();
        horizontalGroup.add(pistolView).fill().uniform();

        leftTable.row();
        leftTable.add(horizontalGroup).growX().center();

        add(leftTable)
                .colspan(1)
                .left()
                .uniformX().fill();

        mainItemsView = new ItemsTableView(localeBundle.get("main.inventory"), titleLabelStyle, subtitleStyle, assetsFinder, skin);
        add(mainItemsView)
                .left()
                .colspan(2)
                .growX();

        OnItemClickListener onItemClickListener = itemModel -> {
            if (itemModel instanceof MedicineModel) {
                MedicineModel model = (MedicineModel) itemModel;
                if (model.getQuantity() > 0) {
                    model.setQuantity(model.getQuantity() - 1);
                    playerSystem.getHealthComponent().treat((MedicineModel) itemModel);
                    soundsSystem.playSound(assetManager.get("audio/sounds/person/medicine.ogg"));
                }
                update();
            } else if (itemModel instanceof WearableModel) {
                dataRepository.getStoryDataModel().setCurrentWearable((WearableModel) itemModel);
                dataRepository.updateDataModel();
                update();
                soundsSystem.playSound(assetManager.get("audio/sounds/person/equip.ogg"));
            }
        };
        mainItemsView.setOnClickListener(onItemClickListener);

        setBackground(Utils.getColoredDrawable(1, 1, Colors.backgroundColor));
        setTouchable(Touchable.enabled);

        dataRepository.addPropertyChangeListener(propertyChangeEvent -> {
            if (propertyChangeEvent.getPropertyName().equals("storyData")) {
                update();
            }
        });
    }

    public void update() {
        StoryDataModel dataModel = dataRepository.getStoryDataModel();
        List<ItemModel> models = dataModel.getAllItems();

        mainItemsView.update(models);

        infoLabel.setText(localeBundle.get("user.info", dataModel.getMoney(), dataModel.getXp()));
        updateWearableInfo();
    }

    public void updateWearableInfo() {
        ArmorModel armorModel = (ArmorModel) dataRepository.getStoryDataModel().getEquippedWearable(ItemType.ARMOR);
        armorView.setWearableModel(armorModel);
        WeaponModel weaponModel = (WeaponModel) dataRepository.getStoryDataModel().getEquippedWearable(ItemType.RIFLE);
        rifleView.setWearableModel(weaponModel);
        weaponModel = (WeaponModel) dataRepository.getStoryDataModel().getEquippedWearable(ItemType.PISTOL);
        pistolView.setWearableModel(weaponModel);
    }

}

File: core/src/net/artux/pda/map/ui/BackpackMenu.kt
package net.artux.pda.map.ui

import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.Touchable
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.VerticalGroup
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import kotlinx.coroutines.runBlocking
import net.artux.engine.utils.LocaleBundle
import net.artux.pda.map.DataRepository
import net.artux.pda.map.di.scope.PerGameMap
import net.artux.pda.map.engine.AssetsFinder
import net.artux.pda.map.engine.systems.SoundsSystem
import net.artux.pda.map.engine.systems.player.PlayerSystem
import net.artux.pda.map.ui.blocks.SlotTextButton
import net.artux.pda.map.ui.view.DetailItemView
import net.artux.pda.map.ui.view.HUD
import net.artux.pda.map.ui.view.ItemsTableView
import net.artux.pda.map.ui.view.OnItemClickListener
import net.artux.pda.map.ui.view.bars.Utils
import net.artux.pda.map.utils.Colors
import net.artux.pda.map.utils.PlatformInterface
import net.artux.pda.model.items.*
import javax.inject.Inject

@PerGameMap
class BackpackMenu @Inject constructor(
    userInterface: UserInterface?,
    platformInterface: PlatformInterface,
    hud: HUD,
    textButton: SlotTextButton,
    adButton: SlotTextButton,
    assetsFinder: AssetsFinder,
    playerSystem: PlayerSystem,
    private val localeBundle: LocaleBundle,
    private val dataRepository: DataRepository,
    skin: Skin?,
    soundsSystem: SoundsSystem
) : Table() {
    private lateinit var assetManager: AssetManager
    private lateinit var fontManager: FontManager
    private lateinit var infoLabel: Label
    private lateinit var mainItemsView: ItemsTableView
    private lateinit var armorView: DetailItemView
    private lateinit var rifleView: DetailItemView
    private lateinit var pistolView: DetailItemView
    fun update() {
        val dataModel = dataRepository.storyDataModelFlow.value
        val models = dataModel.allItems
        mainItemsView.update(models)
        infoLabel.setText(localeBundle["user.info", dataModel.money, dataModel.xp])
        updateWearableInfo()
    }

    fun updateWearableInfo() {
        val currentData = dataRepository.storyDataModelFlow.value

        val armorModel = currentData.getEquippedWearable(ItemType.ARMOR) as ArmorModel
        armorView.setWearableModel(armorModel)

        var weaponModel = currentData.getEquippedWearable(ItemType.RIFLE) as WeaponModel
        rifleView.setWearableModel(weaponModel)

        weaponModel = currentData.getEquippedWearable(ItemType.PISTOL) as WeaponModel
        pistolView.setWearableModel(weaponModel)
    }

    init {
        assetManager = assetsFinder.manager
        top()
        setFillParent(true)
        defaults()
            .pad(10f)
            .space(20f)
        defaults().fill()
        fontManager = assetsFinder.fontManager
        val titleLabelStyle = fontManager.getLabelStyle(38, Color.WHITE)
        val subtitleStyle = fontManager.getLabelStyle(30, Color.GRAY)
        val leftTable = Table()
        leftTable.top().left()
        textButton.setText(localeBundle["main.close"])
        textButton.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                super.clicked(event, x, y)
                remove()
            }
        })
        adButton.setText(localeBundle["main.ad.rewarded"])
        adButton.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent, x: Float, y: Float) {
                platformInterface.rewardedAd()
            }
        })
        infoLabel = Label("", subtitleStyle)
        val verticalGroup = VerticalGroup()
        var buttonsTable = Table()
        buttonsTable.center()
        buttonsTable.defaults().center().space(20f)
        buttonsTable.add(textButton).uniformX()
        buttonsTable.add(adButton).uniformX()
        verticalGroup.addActor(buttonsTable)
        buttonsTable = Table()
        buttonsTable.center()
        buttonsTable.defaults().space(20f).center()
        buttonsTable.add(hud)
        buttonsTable.add(infoLabel)
        buttonsTable.row()
        verticalGroup.addActor(buttonsTable)
        leftTable.add(verticalGroup)
            .left().top().row()
        armorView = DetailItemView(null, titleLabelStyle, subtitleStyle, localeBundle, assetManager)
        leftTable.add(armorView)
            .fill()
        val horizontalGroup = Table()
        horizontalGroup.center()
        rifleView = DetailItemView(null, titleLabelStyle, subtitleStyle, localeBundle, assetManager)
        pistolView =
            DetailItemView(null, titleLabelStyle, subtitleStyle, localeBundle, assetManager)
        pistolView.disableDesc()
        rifleView.disableDesc()
        armorView.disableDesc()
        horizontalGroup.add(rifleView).fill().uniform()
        horizontalGroup.add(pistolView).fill().uniform()
        leftTable.row()
        leftTable.add(horizontalGroup).growX().center()
        add(leftTable)
            .colspan(1)
            .left()
            .uniformX().fill()
        mainItemsView = ItemsTableView(
            localeBundle["main.inventory"],
            titleLabelStyle,
            subtitleStyle,
            assetsFinder,
            skin
        )
        add(mainItemsView)
            .left()
            .colspan(2)
            .growX()
        val onItemClickListener = OnItemClickListener { itemModel: ItemModel? ->
            if (itemModel is MedicineModel) {
                val model = itemModel
                if (model.quantity > 0) {
                    model.quantity = model.quantity - 1
                    playerSystem.healthComponent.treat(itemModel as MedicineModel?)
                    soundsSystem.playSound(assetManager.get("audio/sounds/person/medicine.ogg"))
                }
                update()
            } else if (itemModel is WearableModel) {
                val current = dataRepository.storyDataModelFlow.value
                current.setCurrentWearable(itemModel as WearableModel?)
                dataRepository.storyDataModelFlow.value = current
                soundsSystem.playSound(assetManager.get("audio/sounds/person/equip.ogg"))
            }
        }
        mainItemsView.setOnClickListener(onItemClickListener)
        background = Utils.getColoredDrawable(1, 1, Colors.backgroundColor)
        touchable = Touchable.enabled
        runBlocking {
            dataRepository.storyDataModelFlow.collect {
                update()
            }
        }
    }
}